File: libs/statemanager/src/main/kotlin/mono/state/command/CommandEnvironment.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.command

import mono.graphics.geo.DirectedPoint
import mono.graphics.geo.Point
import mono.graphics.geo.Rect
import mono.livedata.LiveData
import mono.shape.ShapeManager
import mono.shape.connector.ShapeConnector
import mono.shape.selection.SelectedShapeManager
import mono.shape.shape.AbstractShape
import mono.shape.shape.Group
import mono.shape.shape.RootGroup
import mono.shapebound.InteractionPoint

/**
 * An interface defines apis for command to interact with the environment.
 */
internal interface CommandEnvironment {
    val shapeManager: ShapeManager

    val editingModeLiveData: LiveData<EditingMode>

    /**
     * The current working parent group, which is the group that is focused, shape actions will be
     * applied to this group.
     *
     * This is similar to the concept of "current directory" in file system.
     */
    val workingParentGroup: Group

    fun replaceRoot(newRoot: RootGroup, newShapeConnector: ShapeConnector)

    fun enterEditingMode()

    fun exitEditingMode(isNewStateAccepted: Boolean)

    fun addShape(shape: AbstractShape?)

    fun removeShape(shape: AbstractShape?)

    fun getShapes(point: Point): Sequence<AbstractShape>

    fun getAllShapesInZone(bound: Rect): Sequence<AbstractShape>

    fun getWindowBound(): Rect

    fun getInteractionPoint(pointPx: Point): InteractionPoint?

    fun updateInteractionBounds()

    fun isPointInInteractionBounds(point: Point): Boolean

    fun setSelectionBound(bound: Rect?)

    val selectedShapesLiveData: LiveData<Set<AbstractShape>>

    fun getSelectedShapes(): Set<AbstractShape>

    fun addSelectedShape(shape: AbstractShape?)

    fun toggleShapeSelection(shape: AbstractShape)

    fun setFocusingShape(shape: AbstractShape?, focusType: SelectedShapeManager.ShapeFocusType)

    fun getFocusingShape(): AbstractShape?

    fun selectAllShapes()

    fun clearSelectedShapes()

    fun getEdgeDirection(point: Point): DirectedPoint.Direction?

    fun toXPx(column: Double): Double
    fun toYPx(row: Double): Double
    fun toWidthPx(width: Double): Double
    fun toHeightPx(height: Double): Double

    class EditingMode private constructor(val isEditing: Boolean, val skippedVersion: Int?) {
        companion object {
            private val EDIT = EditingMode(true, null)
            private val IDLE = EditingMode(false, null)

            fun edit(): EditingMode = EDIT
            fun idle(skippedVersion: Int?): EditingMode =
                if (skippedVersion == null) IDLE else EditingMode(false, skippedVersion)
        }
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/command-environment.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { Flow } from "$libs/flow";
import { Direction, type Point } from "$libs/graphics-geo/point";
import type { Rect } from "$libs/graphics-geo/rect";
import type { InteractionPoint } from "$mono/shape/interaction-bound";
import { type FocusingShape, ShapeFocusType } from "$mono/shape/selected-shape-manager";
import { type Command, ShapeManager } from '$mono/shape/shape-manager';
import { ShapeConnector } from '$mono/shape/connector/shape-connector';
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import type { Group } from "$mono/shape/shape/group";

/**
 * An interface defines apis for command to interact with the environment.
 */
export interface CommandEnvironment {
    // TODO: Remove this property and add api to access apis of shape manager.
    shapeManager: ShapeManager;

    editingModeFlow: Flow<EditingMode>;

    /**
     * The current working parent group, which is the group that is focused, shape actions will be
     * applied to this group.
     *
     * This is similar to the concept of "current directory" in file system.
     */
    workingParentGroup: Group;

    /**
     * The current root group, which is the group that is the root of the shape tree.
     * This is similar to the concept of "root directory" in file system.
     */
    currentRootGroup: Group;

    replaceRoot(newRoot: Group, newShapeConnector: ShapeConnector): void;

    enterEditingMode(): void;

    exitEditingMode(isNewStateAccepted: boolean): void;

    executeShapeManagerCommand(command: Command): void;

    addShape(shape: AbstractShape | null): void;

    removeShape(shape: AbstractShape | null): void;

    getShapes(point: Point): Iterable<AbstractShape>;

    getAllShapesInZone(bound: Rect): Iterable<AbstractShape>;

    getWindowBound(): Rect;

    getInteractionPoint(pointPx: Point): InteractionPoint | null;

    updateInteractionBounds(): void;

    isPointInInteractionBounds(point: Point): boolean;

    setSelectionBound(bound: Rect | null): void;

    selectedShapesFlow: Flow<Set<AbstractShape>>;

    getSelectedShapes(): Set<AbstractShape>;

    addSelectedShape(shape: AbstractShape | null): void;

    toggleShapeSelection(shape: AbstractShape): void;

    setFocusingShape(shape: AbstractShape | null, focusType: ShapeFocusType): void;

    getFocusingShape(): FocusingShape | null;

    selectAllShapes(): void;

    clearSelectedShapes(): void;

    getEdgeDirection(point: Point): Direction | null;

    toXPx(column: number): number;

    toYPx(row: number): number;

    toWidthPx(width: number): number;

    toHeightPx(height: number): number;
}

export class EditingMode {
    private constructor(public isEditing: boolean, public skippedVersion: number | null) {
    }

    private static readonly IDLE = new EditingMode(false, null);
    private static readonly EDIT = new EditingMode(true, null);

    static edit(): EditingMode {
        return this.EDIT;
    }

    static idle(skippedVersion: number | null): EditingMode {
        return skippedVersion === null ? this.IDLE : new EditingMode(false, skippedVersion);
    }
}
